examples/open-web/src/lib.rs

```diff
@@ -0,0 +1,367 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_bindgen::collections::{Map, Vector};
+use near_bindgen::{env, ext_contract, near_bindgen, Promise, PromiseOrValue};
+use serde::{Deserialize, Serialize};
+
+#[global_allocator]
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+type PublicKey = Vec<u8>;
+type AppId = String;
+type Key = String;
+type Value = String;
+type Message = String;
+type AccountId = String;
+
+const GAS: u64 = 1_000_000_000_000_000;
+
+const APP_METHODS: &[u8] = b"set,remove,pull_message,send_message";
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct UserData {
+    /// App specific PublicKey -> App ID prefix.
+    pub app_keys: Map<PublicKey, AppId>,
+
+    /// Number of app keys per app_id.
+    pub num_app_keys: Map<AppId, u64>,
+
+    /// Received messages. Should be a queue, but whatever.
+    pub messages: Map<AppId, Vector<WrappedMessage>>,
+}
+
+#[derive(BorshDeserialize, BorshSerialize, Serialize, Deserialize)]
+pub struct WrappedMessage {
+    pub sender: AccountId,
+    pub message: Message,
+    pub time: u64,
+}
+
+fn verify_app_id(app_id: &AppId) {
+    if app_id.len() < 2 || app_id.len() > 64 {
+        env::panic(b"App ID length should be between 2 and 64 characters");
+    }
+    for c in app_id.bytes() {
+        match c {
+            b'a'..=b'z' => (),
+            b'0'..=b'9' => (),
+            b'-' | b'_' | b'.' => (),
+            _ => env::panic(
+                b"Unsupported character in the app ID. Only allowed to use `-.|` and 0-9 a-z",
+            ),
+        }
+    }
+}
+
+fn app_key(app_id: &AppId, key: &Key) -> Vec<u8> {
+    let mut res = Vec::with_capacity(app_id.len() + key.len() + 1);
```

Comment:
Why is this mutable?

Reply:
Because I modify it below. In rust you can't modify content of immutable collections.